src/main/java/edu/ie3/datamodel/io/source/csv/CsvGraphicSource.java

```diff
@@ -0,0 +1,224 @@
+import edu.ie3.datamodel.io.factory.input.graphics.LineGraphicInputEntityData;
+import edu.ie3.datamodel.io.factory.input.graphics.LineGraphicInputFactory;
+import edu.ie3.datamodel.io.factory.input.graphics.NodeGraphicInputEntityData;
+import edu.ie3.datamodel.io.factory.input.graphics.NodeGraphicInputFactory;
+import edu.ie3.datamodel.io.source.GraphicSource;
+import edu.ie3.datamodel.io.source.RawGridSource;
+import edu.ie3.datamodel.io.source.TypeSource;
+import edu.ie3.datamodel.models.UniqueEntity;
+import edu.ie3.datamodel.models.input.NodeInput;
+import edu.ie3.datamodel.models.input.OperatorInput;
+import edu.ie3.datamodel.models.input.connector.LineInput;
+import edu.ie3.datamodel.models.input.connector.type.LineTypeInput;
+import edu.ie3.datamodel.models.input.container.GraphicElements;
+import edu.ie3.datamodel.models.input.graphics.LineGraphicInput;
+import edu.ie3.datamodel.models.input.graphics.NodeGraphicInput;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Implementation of the {@link GraphicSource} interface to read {@link NodeGraphicInput} and {@link
+ * LineGraphicInput} entities from .csv files
+ *
+ * @version 0.1
+ * @since 08.04.20
+ */
+public class CsvGraphicSource extends CsvDataSource implements GraphicSource {
+
+  // general fields
+  private final TypeSource typeSource;
+  private final RawGridSource rawGridSource;
+
+  // factories
+  private final LineGraphicInputFactory lineGraphicInputFactory;
+  private final NodeGraphicInputFactory nodeGraphicInputFactory;
+
+  public CsvGraphicSource(
+      String csvSep,
+      String folderPath,
+      FileNamingStrategy fileNamingStrategy,
+      TypeSource typeSource,
+      RawGridSource rawGridSource) {
+    super(csvSep, folderPath, fileNamingStrategy);
+    this.typeSource = typeSource;
+    this.rawGridSource = rawGridSource;
+
+    // init factories
+    this.lineGraphicInputFactory = new LineGraphicInputFactory();
+    this.nodeGraphicInputFactory = new NodeGraphicInputFactory();
+  }
+
+  @Override
+  public Optional<GraphicElements> getGraphicElements() {
+
+    // read all needed entities
+    /// start with types and operators
+    Set<OperatorInput> operators = typeSource.getOperators();
+    Set<LineTypeInput> lineTypes = typeSource.getLineTypes();
+
+    Set<NodeInput> nodes = rawGridSource.getNodes(operators);
+    Set<LineInput> lines = rawGridSource.getLines(nodes, lineTypes, operators);
+
+    // start with the entities needed for a GraphicElements entity
+    /// as we want to return a working grid, keep an eye on empty optionals
+    ConcurrentHashMap<Class<? extends UniqueEntity>, LongAdder> nonBuildEntities =
+        new ConcurrentHashMap<>();
+
+    Set<NodeGraphicInput> nodeGraphics =
```

Comment:
Would it make sense to add join this part and the `getNodeGraphicInput(Set)` method, as they only differ in logging? Might also apply to handling of line graphics and - of course - other sources.

Reply:
What do you mean they differ only in logging? Don't they also differ in the way `Optional.empty()` is handled? Because here we count empty optionals and filter them out afterwards. We don't do this in `getNodeGraphicInput(Set)`?!